zkml/src/model/mod.rs

```diff
@@ -169,61 +169,111 @@ where
         }
     }
 
-    /// Add a re-quantization node to the model after the node with id `input_node_id`
-    pub(crate) fn add_requant_node(
+    /// Add re-quantization nodes to the model after the node with id `input_node_id`
+    /// It creates as many requant layers as there are output wires of the input node
+    pub(crate) fn add_requant_nodes(
         &mut self,
-        requant: Requant,
+        requants: Vec<Requant>,
         input_node_id: NodeId,
-    ) -> anyhow::Result<NodeId> {
+    ) -> anyhow::Result<Vec<NodeId>> {
         let input_node = self
             .nodes
-            .get_mut(&input_node_id)
+            .get(&input_node_id)
             .ok_or(anyhow!("Node {input_node_id} not found in the model"))?;
         let num_outputs = input_node.outputs.len();
-        let requant_node = Node::new_with_outputs(
-            (0..num_outputs)
-                .map(|i| Edge {
-                    node: Some(input_node_id),
-                    index: i,
-                })
-                .collect(),
-            Layer::Requant(requant),
-            input_node.outputs.clone(), // copy output wires of `input_node` to requant node
-        );
-        // remove edges from outputs of `input_node`
-        input_node.outputs = vec![Default::default(); num_outputs];
-        let requant_id = self.add_node(requant_node)?;
+        // we want to create new requant nodes for each output of the input node. That means we need to
+        // create one output edge from input_node to new requant_node and need to copy the associated output wire
+        let requant_nodes = input_node
+            .outputs
+            .iter()
+            .zip(requants.into_iter())
+            .map(|(wire, requant)| {
+                // INPUT EDGES: we look at the output nodes designated by this wire, and for each such node, we look at all the input edges that
+                // come from `input_node_id`. For those, we just copy the index of the wire.
```

Comment:
Isn't this overcomplicated? Here I think we just need to say that the i-th output of `input_node` would be linked to the i-th requant layer, no?